### PR TITLE
Reuse EnumBase for interpreter's Builtin enum

### PR DESCRIPTION
**Pull request by jonmeow**
This was bugging me after I saw all the strings; it feels like this is why we have EnumBase on the toolchain side.

I've included the move of EnumBase to //common because I figured it's reasonable to evaluate together; if we don't want EnumBase in this case, it doesn't make sense to move.

### CODE REVIEW

File: explorer/interpreter/builtins.h [hunk 1]
```diff
@@ -18,145 +19,42 @@
...
+};
+
+class Builtin : public CARBON_ENUM_BASE(Builtin) {
+ public:
+#define CARBON_BUILTIN(Name) CARBON_ENUM_CONSTANT_DECLARATION(Name)
+#include "explorer/interpreter/builtins.def"
+
+  static const int NumBuiltins;
+
+  // Support conversion to and from an int for SemanticNode storage.
```
Comment (zygoloid) [inline explorer/interpreter/builtins.h]: Copy-paste? We don't have SemanticNodes in these parts :)

**Review by zygoloid — APPROVED** [review]
Nice, thanks!

**Review by chandlerc — APPROVED** [review]
(LG to me too, but this is mostly an orthogonal comment)

Would folks be OK consolidating everything into a single top-level `//common`? Not sure that having a separate one in `//toolchain` is pulling its weight.

File: explorer/interpreter/builtins.h (hunk 1 above)
Comment (jonmeow) [reply explorer/interpreter/builtins.h]:
```suggestion
  // Support conversion to and from an int for array indexing.
```

Fixing. :)

**Comment by jonmeow** [comment]
> (LG to me too, but this is mostly an orthogonal comment)
> 
> Would folks be OK consolidating everything into a single top-level `//common`? Not sure that having a separate one in `//toolchain` is pulling its weight.

Here's what that would move:

- explorer
  - arena.h for systems that want to have a leaky approach to memory management.
  - error_builders.h for combining ErrorBuilder with source_location.h
  - nonnull.h for pointer tracking (which we've discussed removing from explorer; although there's been no firm decision)
  - source_location.h for source locations that have no column information.
- toolchain
  - index_base.h for toolchain's int-heavy indexing approach
  - yaml_test_helpers.h which toolchain uses mainly to validate that output is valid yaml (in theory, we could replace this with a simpler validation approach since toiolchain doesn't use it much beyond that).

While sharing EnumBase with explorer was an original motivating factor for its creation (I mentioned this on #2453), I wouldn't expect that to happen with other files. If you still want this done, please let me know and I'll do it.

**Pull request merged**